W5500.device.nut

```diff
@@ -0,0 +1,2538 @@
+    // ***************************************************************************
+    function init(cb) {
+
+        // Configure interrupts
+        setInterrupt(W5500_CONFLICT_INT_TYPE);
+        clearInterrupt();
+        clearSocketInterrupts();
+        _interruptPin.configure(DIGITAL_IN_PULLUP, handleInterrupt.bindenv(this));
+
+        // Set the default mac address
+        setSourceHWAddr(imp.getmacaddress(), true);
+
+        // Set the defaults
+        setNumberOfAvailableSockets(TOTAL_SUPPORTED_SOCKETS);
+
+        // Done
+        imp.wakeup(0, cb);
+
+    }
+
+
+    // ***************************************************************************
+    // forceCloseAllSockets, note this is blocking for 1.1s+
+    // Returns: this
+    // Parameters:
+    //          none
+    // **************************************************************************
+    function forceCloseAllSockets() {
+        _maxNoOfSockets = getTotalSupportedSockets();
+        for (local socket = 0; socket < _maxNoOfSockets; socket++) {
+            sendSocketCommand(socket, W5500_SOCKET_DISCONNECT);
+        }
+        imp.sleep(1);
+        for (local socket = 0; socket < _maxNoOfSockets; socket++) {
+            sendSocketCommand(socket, W5500_SOCKET_CLOSE);
+        }
+        imp.sleep(0.1);
+    }
+
+
+    // ***************************************************************************
+    // getPhysicalLinkStatus
+    // Returns: true or false
+    // Parameters:
+    //      none
+    // ***************************************************************************
+    function getPhysicalLinkStatus() {
+        local status = readReg(W5500_PHYSICAL_CONFIG, W5500_COMMON_REGISTER);
+        return (status & 0x01) == 0x01;
+    }
+
+
+    // ***************************************************************************
+    // openConnection
+    // Returns: this
+    // Parameters:
+    //      destIP - the ip address of the destination
+    //      destPort - the port of the destination 
+    //      mode - TCP or UDP
+    //      sourcePort(optional) - the port to send from
+    //      cb - function to be called when connection successfully established 
+    // ****************************************************************************
+    function openConnection(destIP, destPort, mode, sourcePort = null, cb = null) {
+
+        // shuffle oarameters around
+        if (typeof sourcePort == "function") {
+            cb = sourcePort;
+            sourcePort = null;
+        }
+
+        // check for available socket
+        if (_availableSockets.len() == 0) {
+            if (cb) return cb(W5500_ERR_CANNOT_CONNECT_SOCKETS_IN_USE, null);
+            else throw W5500_ERR_CANNOT_CONNECT_SOCKETS_IN_USE;
+        }
+
+        // Create the connection object and assign it to a spare socket
+        local socket = _availableSockets.pop();
+        _connections[socket] <- W5500.Connection(this, socket, destIP, destPort, mode);
```

Comment:
We shouldn't be a) polling here and b) do in in a blocking way so that the imp can't process any external events and requests. Just doesn't seem the right thing to do.
Interrupts instead of polling might probably be a better way to address this.